src/Components/Web/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides common functionality for virtualized lists.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public abstract class VirtualizeBase<TItem> : ComponentBase, IVirtualizeJsCallbacks, IAsyncDisposable
+    {
+        private VirtualizeJsInterop _jsInterop = default!;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <summary>
+        /// Gets the number of items rendered before the visible part of the container.
+        /// </summary>
+        protected int ItemsBefore { get; private set; }
+
+        /// <summary>
+        /// Gets the number of items the container can visibly fit.
+        /// </summary>
+        protected int VisibleItemCapacity { get; private set; }
+
+        /// <summary>
+        /// Gets the number of items currently visible within the container.
+        /// </summary>
+        protected int ItemsVisible { get; private set; }
+
+        /// <summary>
+        /// Gets the total number of items.
+        /// </summary>
+        protected abstract int ItemCount { get; }
+
+        /// <summary>
+        /// Gets the size of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Renders the visible items in the list.
+        /// </summary>
+        /// <param name="builder">A <see cref="RenderTreeBuilder"/> to receive the render output.</param>
+        protected abstract void RenderItems(RenderTreeBuilder builder);
+
+        /// <inheritdoc />
+        public override async Task SetParametersAsync(ParameterView parameters)
+        {
+            await base.SetParametersAsync(parameters);
+
+            _jsInterop ??= new VirtualizeJsInterop(this, JSRuntime);
```

Comment:
It won't make much difference, but since we don't use `_jsInterop` until `OnAfterRenderAsync`, could this line be moved down to (say) just before L77?